# Exploratory Analysis

How car sales work in ebay

In [13]:
import pandas as pd
import numpy as np
import re

In [14]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [15]:
autos.head(1)

dateCrawled                              name  seller offerType  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat   Angebot   

    price   abtest vehicleType  yearOfRegistration  gearbox  powerPS   model  \
0  $5,000  control         bus                2004  manuell      158  andere   

    odometer  monthOfRegistration fuelType    brand notRepairedDamage  \
0  150,000km                    3      lpg  peugeot              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-26 00:00:00             0       79588  2016-04-06 06:45:54

From this point, the created dataframe is a german list of cars sold in eBay, the 'name' column is an aggregatted set of relevant info hard to extract and make conclusions. Most of columns are objects and all of them have relevant not null data.

In [16]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [17]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_p_s', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

From now on, the columns have snakecase names, that easier to read and deal with Python code.

In [18]:
autos.describe()

year_of_registration     power_p_s  month_of_registration  \
count          50000.000000  50000.000000           50000.000000   
mean            2005.073280    116.355920               5.723360   
std              105.712813    209.216627               3.711984   
min             1000.000000      0.000000               0.000000   
25%             1999.000000     70.000000               3.000000   
50%             2003.000000    105.000000               6.000000   
75%             2008.000000    150.000000               9.000000   
max             9999.000000  17700.000000              12.000000   

       nr_of_pictures   postal_code  
count         50000.0  50000.000000  
mean              0.0  50813.627300  
std               0.0  25779.747957  
min               0.0   1067.000000  
25%               0.0  30451.000000  
50%               0.0  49577.000000  
75%               0.0  71540.000000  
max               0.0  99998.000000

nr_of_pictures column have all columns with 0 value
power_p_s and year_of_registration have weird max numbers
price and odometer should be numbers instead of strings

In [19]:
autos['price'] = (autos['price']
                  .str.replace(r'[$|,]', '')
                  .astype(float)
                 )

autos['odometer'] = (autos['odometer']
                     .str.replace(r'[km|,]', '')
                     .astype(int)
                    )
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

In [20]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

## Removing Outliers

In [21]:
autos['price'].unique().shape

(2357,)

In [22]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [23]:
outliers = autos['price'].value_counts().sort_index(ascending=True)[50:180]
outliers

50.0      49
55.0       2
59.0       1
60.0       9
65.0       5
66.0       1
70.0      10
75.0       5
79.0       1
80.0      15
89.0       1
90.0       5
99.0      19
100.0    134
110.0      3
111.0      2
115.0      2
117.0      1
120.0     39
122.0      1
125.0      8
129.0      1
130.0     15
135.0      1
139.0      1
140.0      9
145.0      2
149.0      7
150.0    224
156.0      2
160.0      8
170.0      7
173.0      1
175.0     12
179.0      1
180.0     35
Name: price, dtype: int64

Any value below 100 will be discarted. 
These values do not represent any viable car's price
100 is also the first number with more than 100 occurences

In [24]:
outliers = autos['price'].value_counts().sort_index(ascending=False).head(15)
outliers

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price, dtype: int64

Any value above 350000 will be discarted. These values are at least 2,5 times bigger than any  other in list.

In [25]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

There's no outliers in odometer_km column

In [26]:
autos = autos.loc[autos['price'].between(100, 350000)]
autos['price'].describe()

count     48224.000000
mean       5930.371433
std        9078.372762
min         100.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

In [27]:
autos.describe()

price  year_of_registration     power_p_s    odometer_km  \
count   48224.000000          48224.000000  48224.000000   48224.000000   
mean     5930.371433           2004.730964    117.677609  125919.148142   
std      9078.372762             87.897388    201.206304   39543.339640   
min       100.000000           1000.000000      0.000000    5000.000000   
25%      1250.000000           1999.000000     73.000000  125000.000000   
50%      3000.000000           2004.000000    107.000000  150000.000000   
75%      7499.000000           2008.000000    150.000000  150000.000000   
max    350000.000000           9999.000000  17700.000000  150000.000000   

       month_of_registration  nr_of_pictures   postal_code  
count           48224.000000         48224.0  48224.000000  
mean                5.801634             0.0  50987.919729  
std                 3.676976             0.0  25737.119986  
min                 0.000000             0.0   1067.000000  
25%                 3.000000             0.0  30823.000000  
50%                 6.000000             0.0  49716.000000  
75%                 9.000000             0.0  71666.750000  
max                12.000000             0.0  99998.000000

Assuming that in average each car runs 10.000km per year, it is possible to infer year_of_registration for the invalid cells.

## Treating Dates

In [28]:
autos[['date_crawled', 'ad_created', 'last_seen']].describe()

date_crawled           ad_created            last_seen
count                 48224                48224                48224
unique                46571                   76                38232
top     2016-03-19 17:36:18  2016-04-03 00:00:00  2016-04-07 06:17:27
freq                      3                 1874                    8

In [29]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025361
2016-03-06    0.014039
2016-03-07    0.036061
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-10    0.032287
2016-03-11    0.032598
2016-03-12    0.036911
2016-03-13    0.015677
2016-03-14    0.036662
2016-03-15    0.034319
2016-03-16    0.029467
2016-03-17    0.031499
2016-03-18    0.012898
2016-03-19    0.034734
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-22    0.032888
2016-03-23    0.032287
2016-03-24    0.029446
2016-03-25    0.031499
2016-03-26    0.032308
2016-03-27    0.031126
2016-03-28    0.034962
2016-03-29    0.034112
2016-03-30    0.033738
2016-03-31    0.031851
2016-04-01    0.033697
2016-04-02    0.035605
2016-04-03    0.038611
2016-04-04    0.036538
2016-04-05    0.013064
2016-04-06    0.003173
2016-04-07    0.001389
Name: date_crawled, dtype: float64

In [30]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033096
2016-03-10    0.031997
2016-03-11    0.032909
2016-03-12    0.036745
2016-03-13    0.017045
2016-03-14    0.035294
2016-03-15    0.034049
2016-03-16    0.029964
2016-03-17    0.031167
2016-03-18    0.013582
2016-03-19    0.033614
2016-03-20    0.037865
2016-03-21 

After the crawler implementation, ad_created registration grew 20 times.

In [31]:
autos['last_seen'].str[:10].value_counts(dropna=False).sort_index()

2016-03-05       52
2016-03-06      208
2016-03-07      262
2016-03-08      353
2016-03-09      462
2016-03-10      513
2016-03-11      598
2016-03-12     1147
2016-03-13      428
2016-03-14      609
2016-03-15      765
2016-03-16      793
2016-03-17     1355
2016-03-18      353
2016-03-19      760
2016-03-20      996
2016-03-21      991
2016-03-22     1030
2016-03-23      896
2016-03-24      953
2016-03-25      921
2016-03-26      804
2016-03-27      750
2016-03-28     1005
2016-03-29     1075
2016-03-30     1191
2016-03-31     1149
2016-04-01     1102
2016-04-02     1200
2016-04-03     1212
2016-04-04     1183
2016-04-05     6031
2016-04-06    10704
2016-04-07     6373
Name: last_seen, dtype: int64

The dates between '2016-04-05' and '2016-04-07' had significant raise. Eighter the crawler and ad creation had the same change.

In [32]:
autos['year_of_registration'].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

The year_of_registration column have some inconsistent values that have to be cleaned like year 1000 and 9999

In [33]:
autos['year_of_registration'].value_counts(normalize=True, dropna=False).sort_index()

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000041
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000021
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
1955    0.000041
1956    0.000083
1957    0.000041
1958    0.000083
1959    0.000124
1960    0.000456
1961    0.000124
1962    0.000083
1963    0.000166
1964    0.000249
          ...   
2000    0.064366
2001    0.054516
2002    0.051364
2003    0.055844
2004    0.055968
2005    0.060364
2006    0.055325
2007    0.047134
2008    0.045828
2009    0.043132
2010    0.032909
2011    0.033552
2012    0.027123
2013    0.016610
2014    0.013728
2015    0.007880
2016    0.024925
2017    0.028679
2018    0.009705
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000062
5911    0.000021
6200    0.000021
8888    0.0000

In [34]:
autos.loc[~autos['year_of_registration'].between(1900, 2016), 'year_of_registration'].count()

1872

In this analysis it is considered just modern model cars.
Despite of removing wrong data, cars older than 1900 will not be considered.

Before 1900, car's architecture were different from today. Thus, for eBay car section this type of product may have different consumers, prices and specifications.

This dataframe does not contain data input after 2016, so, car's model can't be for later then 2016.

## Remove Outliers

In [35]:
autos = autos.loc[autos['year_of_registration'].between(1900, 2016)]

In [36]:
autos['year_of_registration'].value_counts(normalize=True, dropna= False).sort_values(ascending=False)

2000    0.066966
2005    0.062802
1999    0.062112
2004    0.058228
2003    0.058099
2006    0.057560
2001    0.056718
2002    0.053439
1998    0.050483
2007    0.049038
2008    0.047679
2009    0.044874
1997    0.041530
2011    0.034907
2010    0.034238
1996    0.029233
2012    0.028219
2016    0.025932
1995    0.025738
2013    0.017281
2014    0.014282
1994    0.013505
1993    0.009061
2015    0.008198
1992    0.007918
1991    0.007292
1990    0.007163
1989    0.003689
1988    0.002869
1985    0.002006
          ...   
1960    0.000475
1976    0.000453
1969    0.000410
1975    0.000388
1965    0.000367
1964    0.000259
1963    0.000173
1961    0.000129
1959    0.000129
1937    0.000086
1958    0.000086
1962    0.000086
1956    0.000086
1955    0.000043
1954    0.000043
1957    0.000043
1951    0.000043
1910    0.000043
1934    0.000043
1941    0.000043
1948    0.000022
1953    0.000022
1939    0.000022
1938    0.000022
1943    0.000022
1931    0.000022
1950    0.000022
1929    0.0000

Half of the models have registration_year between 1999 - 2009.

## Selection of most relevant Brands

In [37]:
eligible_brands = (autos['brand']
                   .value_counts(normalize=True, dropna=False) > 0.005
                  ).value_counts()

eligible_brands

True     27
False    13
Name: brand, dtype: int64

27 Brands have at least 5% share of the car's catalog were considered

## Brand's Prices comparison

In [38]:
brands = (autos['brand']
          .value_counts(normalize=True)
          .sort_values(ascending=False).head(27)
         ).index

brand_mean_prices = {}
brand_car_miles = {}

for b in brands:
    brand_mean_prices[b] = autos.loc[autos['brand'] == b, 'price'].mean()
    brand_car_miles[b] = autos.loc[autos['brand'] == b, 'odometer_km'].mean()

brand_mean_prices

bmp_series = pd.Series(brand_mean_prices)
bcm_series = pd.Series(brand_car_miles)

b_df = pd.DataFrame(data={'mean_price':bmp_series,'mean_miles':bcm_series})
b_df.sort_values(by='mean_miles', ascending=False)

mean_price     mean_miles
volvo            4993.208038  138581.560284
bmw              8381.677306  132695.320149
alfa_romeo       4100.915858  131747.572816
mercedes_benz    8672.654241  131025.669643
opel             3005.496077  129384.429692
audi             9380.718548  129245.400298
volkswagen       5436.950097  128799.877539
renault          2496.070577  128281.393217
peugeot          3113.860549  127127.890173
mitsubishi       3439.102902  127097.625330
mazda            4129.774788  124553.824363
ford             3779.269076  124277.108434
honda            4119.109589  122493.150685
seat             4433.419622  121536.643026
citroen          3796.262673  119554.531490
nissan           4756.659634  118326.300985
fiat             2836.873631  116950.294861
toyota           5167.091062  115944.350759
kia              6018.442073  112301.829268
skoda            6409.609724  110906.701708
suzuki           4126.341818  108163.636364
hyundai          5411.075431  106885.775862
chevrolet        6759.885932  100266.159696
smart            3596.402736   99734.042553
porsche         46788.444444   98476.702509
sonstige_autos  12784.556561   90395.927602
mini            10639.450980   88308.823529

Cars from 'other' brands have the most highest mean price, it can represent old brands or office brands. 

Despite this, the most expensive are: Audi, Mercedes, BMW and Mini.
On the other hand, Renault, Smart, Ford, Fiat and Citroen have the lowest mean prices

From the most expensive companies, BMW has the highest miles' mean.
Opel and Volvo are the less expensive cars that have the biggest miles' average

## Extra analysis

In [39]:
vehicle_dict= {
    'bus': 'bus',
    'limousine': 'limousine',
    'kleinwagen': 'hatch',
    'kombi': 'van',
    'nan': np.nan,
    'coupe': 'sedan',
    'suv': 'suv',
    'cabrio': 'conversible',
    'andere': 'other'
}

autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_dict)

In [40]:
gearbox_dict = {
    'manuell':'manual',
    'automatik':'automatic',
    'nan':np.nan
}

autos['gearbox'] = autos['gearbox'].map(gearbox_dict)

In [41]:
fuel_dict = {
    'lpg':'lpg',
    'benzin':'petrol',
    'diesel':'diesel',
    'nan':np.nan,
    'cng':'cng',
    'hybrid':'hybrid',
    'elektro':'electric',
    'andere':'other'
}

autos['fuel_type'] = autos['fuel_type'].map(fuel_dict)

In [42]:
repair_dict = {
    'nein':np.nan,
    'nan':np.nan,
    'ja':'yes'
}

autos['not_repaired_damage'] = autos['not_repaired_damage'].map(repair_dict)

In [43]:
autos['date_crawled'] = autos['date_crawled'].str.replace('-','').str.split(expand=True).iloc[:,0]
autos['ad_created'] = autos['ad_created'].str.replace('-','').str.split(expand=True).iloc[:,0]
autos['last_seen'] = autos['last_seen'].str.replace('-','').str.split(expand=True).iloc[:,0]

In [44]:
autos['motor_size'] = autos['name']

autos['motor_size'] = (autos['motor_size']
                      .str.replace('_', ' ')
                      .str.replace(r'([0-9]{2,})', '')
                      .str.extract('([0-9]\.[0-9])')
                      .astype(float)
                      )

autos.loc[autos['motor_size'] > 8.5] = np.nan

In [60]:
autos

date_crawled                                               name  seller  \
0         20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1         20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2         20160326                         Volkswagen_Golf_1.6_United  privat   
3         20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4         20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   
5         20160321  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...  privat   
6         20160320  VW_Golf_III_GT_Special_Electronic_Green_Metall...  privat   
7         20160316                               Golf_IV_1.9_TDI_90PS  privat   
8         20160322                                         Seat_Arosa  privat   
9         20160316          Renault_Megane_Scenic_1.6e_RT_Klimaanlage  privat   
11        20160316                         Mercedes_A140_Motorschaden  privat   
12        20160331  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...  privat   
13        20160323                                 Audi_A3_1.6_tuning  privat   
14        20160323  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...  privat   
15        20160401        Corvette_C3_Coupe_T_Top_Crossfire_Injection  privat   
16        20160316                                Opel_Vectra_B_Kombi  privat   
17        20160329                          Volkswagen_Scirocco_2_G60  privat   
18        20160326               Verkaufen_mein_bmw_e36_320_i_touring  privat   
19        20160317        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018  privat   
20        20160305  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...  privat   
21        20160306                      Porsche_911_Carrera_4S_Cabrio  privat   
22        20160328                               MINI_Cooper_S_Cabrio  privat   
23        20160310   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima  privat   
24        20160403                         BMW_535i_xDrive_Sport_Aut.  privat   
26        20160403                                Volkswagen_Polo_Fox  privat   
28        20160319                                      MINI_Cooper_D  privat   
29        20160402  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...  privat   
31        20160314                          Noch_gut_erhaltenen_C_320  privat   
32        20160320                               Corsa_mit_TÜV_5.2016  privat   
33        20160315  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...  privat   
...            ...                                                ...     ...   
49968     20160401                  Mercedes_Benz_190_D_2.5_Automatik  privat   
49969     20160317             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ  privat   
49970     20160321  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...  privat   
49971     20160329                                         W.Lupo_1.0  privat   
49972     20160326          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.  privat   
49973     20160327                   Mercedes_Benz_SLK_200_Kompressor  privat   
49975     20160327          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort  privat   
49976     20160319  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...  privat   
49977     20160331                        Mercedes_Benz_C200_Cdi_W203  privat   
49978     20160404                        Mercedes_Benz_E_200_Classic  privat   
49979     20160320                      Volkswagen_Polo_1.6_TDI_Style  privat   
49980     20160312                            Ford_Escort_Turnier_16V  privat   
49981     20160315             Opel_Astra_Kombi_mit_Anhaengerkupplung  privat   
49982     20160329            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm  privat   
49983     20160306                                      Ford_focus_99  privat   
49985     20160402                          Verkaufe_meinen_vw_vento!  privat   
49986     20160404  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...  privat   
49987     20160322  Audi_A3_Limousine

## Look for brand / model common combinations

In [66]:
brands = autos['brand'].unique()
types = autos['vehicle_type'].unique()

brand_vehicles_per_type = {}
for brand in brands:
    for type in types:
        autos.loc['']

array(['bus', 'limousine', 'hatch', 'van', nan, 'sedan', 'suv',
       'conversible', 'other'], dtype=object)